diff --git a/dbms/src/DataStreams/ReplacingSortedBlockInputStream.h b/dbms/src/DataStreams/ReplacingSortedBlockInputStream.h
index b8592a0..26e473e 100644
--- a/dbms/src/DataStreams/ReplacingSortedBlockInputStream.h
+++ b/dbms/src/DataStreams/ReplacingSortedBlockInputStream.h
@@ -16,9 +16,9 @@ class ReplacingSortedBlockInputStream : public MergingSortedBlockInputStream
 {
 public:
     ReplacingSortedBlockInputStream(BlockInputStreams inputs_, const SortDescription & description_,
-        const String & version_column_, size_t max_block_size_, WriteBuffer * out_row_sources_buf_ = nullptr)
+        const String & version_column_, size_t max_block_size_, WriteBuffer * out_row_sources_buf_ = nullptr, bool skip_single_child_ = true)
         : MergingSortedBlockInputStream(inputs_, description_, max_block_size_, 0, out_row_sources_buf_),
-        version_column(version_column_)
+        version_column(version_column_), skip_single_child(skip_single_child_)
     {
     }
 
@@ -30,6 +30,7 @@ protected:
 
 private:
     String version_column;
+    bool skip_single_child;
     ssize_t version_column_number = -1;
 
     Logger * log = &Logger::get("ReplacingSortedBlockInputStream");
