diff --git a/dbms/src/Interpreters/Aggregator.cpp b/dbms/src/Interpreters/Aggregator.cpp
index 0b458de253c6..2f15b201725c 100644
--- a/dbms/src/Interpreters/Aggregator.cpp
+++ b/dbms/src/Interpreters/Aggregator.cpp
@@ -441,6 +441,8 @@ AggregatedDataVariants::Type Aggregator::chooseAggregationMethod()
             return AggregatedDataVariants::Type::key64;
         if (size_of_field == 16)
             return AggregatedDataVariants::Type::keys128;
+        if (size_of_field == sizeof(DecimalValue))
+            return AggregatedDataVariants::Type::key_decimal;
         throw Exception("Logical error: numeric column has sizeOfField not in 1, 2, 4, 8, 16.", ErrorCodes::LOGICAL_ERROR);
     }
 
@@ -577,6 +579,8 @@ void NO_INLINE Aggregator::executeImplCase(
                     prev_key = key;
             }
 
+//            if constexpr(std::is_same_v<Method::Key, DB::DecimalValue>)
+ //               std::cout<<(int)key.value<<std::endl;
             method.data.emplace(key, it, inserted);
         }
         else
