diff --git a/dbms/src/Columns/ColumnVector.h b/dbms/src/Columns/ColumnVector.h
index 5ce33e820282..48329873d033 100644
--- a/dbms/src/Columns/ColumnVector.h
+++ b/dbms/src/Columns/ColumnVector.h
@@ -226,15 +226,9 @@ public:
 
     UInt64 get64(size_t n) const override;
 
-    UInt64 getUInt(size_t n) const override
-    {
-        return UInt64(data[n]);
-    }
+    UInt64 getUInt(size_t n) const override;
 
-    Int64 getInt(size_t n) const override
-    {
-        return Int64(data[n]);
-    }
+    Int64 getInt(size_t n) const override;
 
     void insert(const Field & x) override
     {
