diff --git a/dbms/src/Storages/MergeTree/MergeTreeSettings.h b/dbms/src/Storages/MergeTree/MergeTreeSettings.h
index f91ca880a661..377b4fd3ca4d 100644
--- a/dbms/src/Storages/MergeTree/MergeTreeSettings.h
+++ b/dbms/src/Storages/MergeTree/MergeTreeSettings.h
@@ -21,6 +21,9 @@ struct MergeTreeSettings
     /** How many rows correspond to one primary key value. */                                                 \
     M(SettingUInt64, index_granularity, 8192)                                                                 \
                                                                                                               \
+    /** MutableMergeTree partition number. */                                                                 \
+    M(SettingUInt64, mutable_mergetree_partition_number, 16)                                                  \
+                                                                                                              \
     /** Merge settings. */                                                                                    \
                                                                                                               \
     /** Maximum in total size of parts to merge, when there are maximum (minimum) free threads                \
