diff --git a/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.h b/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.h
index 1b3caeeb2..28633dbed 100644
--- a/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.h
+++ b/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.h
@@ -58,6 +58,17 @@ private:
         const Settings & settings,
         const Context & context) const;
 
+    BlockInputStreams spreadMarkRangesAmongStreamsOnMutableEngine(
+        const RangesInDataParts & parts_with_ranges,
+        const Names & column_names,
+        size_t max_block_size,
+        unsigned num_streams,
+        bool use_uncompressed_cache,
+        ExpressionActionsPtr prewhere_actions,
+        const String & prewhere_column,
+        const Names & virt_columns,
+        const Settings & settings) const;
+
     /// Get the approximate value (bottom estimate - only by full marks) of the number of rows falling under the index.
     size_t getApproximateTotalRowsToRead(
         const MergeTreeData::DataPartsVector & parts,
