diff --git a/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp b/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp
index afc9700..6bf1f1b 100644
--- a/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp
+++ b/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp
@@ -7,10 +7,14 @@
 #include <Storages/MergeTree/MergeTreeReadPool.h>
 #include <Storages/MergeTree/MergeTreeThreadBlockInputStream.h>
 #include <Storages/MergeTree/PKCondition.h>
+
 #include <Parsers/ASTIdentifier.h>
 #include <Parsers/ASTFunction.h>
 #include <Parsers/ASTSampleRatio.h>
 
+#include <DataStreams/DebugPrintBlockInputStream.h>
+#include <DataStreams/DedupSortedBlockInputStream.h>
+
 /// Allow to use __uint128_t as a template parameter for boost::rational.
 // https://stackoverflow.com/questions/41198673/uint128-t-not-working-with-clang-and-libstdc
 #if !defined(__GLIBCXX_BITSIZE_INT_N_0) && defined(__SIZEOF_INT128__)
@@ -44,6 +48,11 @@ namespace std
 #include <DataTypes/DataTypeEnum.h>
 #include <Storages/VirtualColumnUtils.h>
 
+#include <Storages/MutableSupport.h>
+#include <DataStreams/ExtraHeadTailBlockInputStream.h>
+#include <DataStreams/ReplacingDeletingSortedBlockInputStream.h>
+#include <DataStreams/AsynchronousBlockInputStream.h>
+
 
 namespace ProfileEvents
 {
@@ -139,8 +148,6 @@ BlockInputStreams MergeTreeDataSelectExecutor::read(
     const unsigned num_streams,
     Int64 max_block_number_to_read) const
 {
-    size_t part_index = 0;
-
     MergeTreeData::DataPartsVector parts = data.getDataPartsVector();
 
     /// If query contains restrictions on the virtual column `_part` or `_part_index`, select only parts suitable for it.
@@ -546,9 +553,37 @@ BlockInputStreams MergeTreeDataSelectExecutor::read(
     ProfileEvents::increment(ProfileEvents::SelectedRanges, sum_ranges);
     ProfileEvents::increment(ProfileEvents::SelectedMarks, sum_marks);
 
+    bool is_mutable_engine = (data.merging_params.mode == MergeTreeData::MergingParams::Mutable && !select.raw_for_mutable);
+
     BlockInputStreams res;
 
-    if (select.final())
+    if (is_mutable_engine)
+    {
+        column_names_to_read.insert(column_names_to_read.end(), MutableSupport::version_column_name);
+        column_names_to_read.insert(column_names_to_read.end(), MutableSupport::delmark_column_name);
+
+        std::vector<String> add_columns = data.getPrimaryExpression()->getRequiredColumns();
+        column_names_to_read.insert(column_names_to_read.end(), add_columns.begin(), add_columns.end());
+
+        std::sort(column_names_to_read.begin(), column_names_to_read.end());
+        column_names_to_read.erase(std::unique(column_names_to_read.begin(), column_names_to_read.end()), column_names_to_read.end());
+
+        res = spreadMarkRangesAmongStreamsOnMutableEngine(
+            parts,
+            key_condition,
+            column_names_to_read,
+            max_block_size,
+            num_streams,
+            settings.use_uncompressed_cache,
+            prewhere_actions,
+            prewhere_column,
+            virt_column_names,
+            settings);
+
+        if (res.size() == 0)
+            return res;
+    }
+    else if (select.final())
     {
         /// Add columns needed to calculate primary key and the sign.
         std::vector<String> add_columns = data.getPrimaryExpression()->getRequiredColumns();
@@ -743,6 +778,7 @@ BlockInputStreams MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreams(
     return res;
 }
 
+
 BlockInputStreams MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreamsFinal(
     RangesInDataParts && parts,
     const Names & column_names,
@@ -833,6 +869,9 @@ BlockInputStreams MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreamsFinal
 
             case MergeTreeData::MergingParams::Graphite:
                 throw Exception("GraphiteMergeTree doesn't support FINAL", ErrorCodes::LOGICAL_ERROR);
+
+            case MergeTreeData::MergingParams::Mutable:
+                throw Exception("MutableMergeTree doesn't handle here", ErrorCodes::LOGICAL_ERROR);
         }
 
         res.emplace_back(merged);
@@ -842,6 +881,163 @@ BlockInputStreams MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreamsFinal
 }
 
 
+BlockInputStreams MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreamsOnMutableEngine(
+    MergeTreeData::DataPartsVector & parts,
+    const PKCondition & key_condition,
+    const Names & column_names,
+    size_t max_block_size,
+    unsigned num_streams,
+    bool use_uncompressed_cache,
+    ExpressionActionsPtr prewhere_actions,
+    const String & prewhere_column,
+    const Names & virt_columns,
+    const Settings & settings) const
+{
+    LOG_DEBUG(log, "Number of streams: " << num_streams);
+
+    RangesInDataParts parts_with_ranges;
+    size_t sum_marks = 0;
+    getPartsWithRanges(parts, key_condition, settings, parts_with_ranges, &sum_marks);
+
+    const size_t max_marks_to_use_cache =
+        (settings.merge_tree_max_rows_to_use_cache + data.index_granularity - 1) / data.index_granularity;
+    if (sum_marks > max_marks_to_use_cache)
+        use_uncompressed_cache = false;
+
+    std::unordered_map<std::string, RangesInDataParts> partitions;
+    for (size_t part_index = 0; part_index < parts_with_ranges.size(); ++part_index)
+    {
+        RangesInDataPart & part = parts_with_ranges[part_index];
+        auto partition_id = part.data_part->partition.getID(data);
+        if (partitions.find(partition_id) == partitions.end())
+            partitions[partition_id] = RangesInDataParts();
+        partitions[partition_id].emplace_back(part);
+    }
+
+    // TODO: Use one stream if data is small enough
+
+    BlockInputStreams res;
+
+    if (MutableSupport::deduper == MutableSupport::DeduperOriginStreams)
+    {
+        for (auto it = parts_with_ranges.begin(); it != parts_with_ranges.end(); ++it)
+        {
+            RangesInDataPart & part = *it;
+            BlockInputStreamPtr source_stream = std::make_shared<MergeTreeBlockInputStream>(
+                data, part.data_part, max_block_size, settings.preferred_block_size_bytes,
+                settings.preferred_max_column_in_block_size_bytes, column_names, part.ranges, use_uncompressed_cache,
+                prewhere_actions, prewhere_column, true, settings.min_bytes_to_use_direct_io, settings.max_read_buffer_size, true,
+                virt_columns, part.part_index_in_query);
+            // source_stream = std::make_shared<DebugPrintBlockInputStream>(source_stream);
+            source_stream = std::make_shared<ExpressionBlockInputStream>(source_stream, data.getPrimaryExpression());
+            res.emplace_back(source_stream);
+        }
+    }
+    else if (MutableSupport::deduper == MutableSupport::DeduperOriginUnity)
+    {
+        BlockInputStreams merging;
+        for (auto it = parts_with_ranges.begin(); it != parts_with_ranges.end(); ++it)
+        {
+            RangesInDataPart & part = *it;
+            BlockInputStreamPtr source_stream = std::make_shared<MergeTreeBlockInputStream>(
+                data, part.data_part, max_block_size, settings.preferred_block_size_bytes,
+                settings.preferred_max_column_in_block_size_bytes, column_names, part.ranges, use_uncompressed_cache,
+                prewhere_actions, prewhere_column, true, settings.min_bytes_to_use_direct_io, settings.max_read_buffer_size, true,
+                virt_columns, part.part_index_in_query);
+            // source_stream = std::make_shared<DebugPrintBlockInputStream>(source_stream);
+            source_stream = std::make_shared<ExpressionBlockInputStream>(source_stream, data.getPrimaryExpression());
+            merging.emplace_back(source_stream);
+        }
+        res.emplace_back(std::make_shared<ReplacingDeletingSortedBlockInputStream>(merging, data.getSortDescription(),
+            MutableSupport::version_column_name, MutableSupport::delmark_column_name, DEFAULT_MERGE_BLOCK_SIZE, nullptr, false, true));
+    }
+    else if (MutableSupport::deduper == MutableSupport::DeduperReplacingUnity)
+    {
+        BlockInputStreams merging;
+        for (auto it = partitions.begin(); it != partitions.end(); ++it)
+        {
+            RangesInDataParts & parts = it->second;
+
+            for (size_t i = 0; i < parts.size(); ++i)
+            {
+                RangesInDataPart & part = parts[i];
+                BlockInputStreamPtr source_stream = std::make_shared<MergeTreeBlockInputStream>(
+                    data, part.data_part, max_block_size, settings.preferred_block_size_bytes,
+                    settings.preferred_max_column_in_block_size_bytes, column_names, part.ranges, use_uncompressed_cache,
+                    prewhere_actions, prewhere_column, true, settings.min_bytes_to_use_direct_io, settings.max_read_buffer_size, true,
+                    virt_columns, part.part_index_in_query);
+                // source_stream = std::make_shared<DebugPrintBlockInputStream>(source_stream);
+                source_stream = std::make_shared<ExpressionBlockInputStream>(source_stream, data.getPrimaryExpression());
+                merging.emplace_back(source_stream);
+            }
+        }
+        res.emplace_back(std::make_shared<ReplacingDeletingSortedBlockInputStream>(merging, data.getSortDescription(),
+            MutableSupport::version_column_name, MutableSupport::delmark_column_name, DEFAULT_MERGE_BLOCK_SIZE, nullptr, false, true));
+    }
+    else if (MutableSupport::deduper == MutableSupport::DeduperReplacingPartitioning)
+    {
+        for (auto it = partitions.begin(); it != partitions.end(); ++it)
+        {
+            RangesInDataParts & parts = it->second;
+            BlockInputStreams merging;
+
+            for (size_t i = 0; i < parts.size(); ++i)
+            {
+                RangesInDataPart & part = parts[i];
+
+                BlockInputStreamPtr source_stream = std::make_shared<MergeTreeBlockInputStream>(
+                    data, part.data_part, max_block_size, settings.preferred_block_size_bytes,
+                    settings.preferred_max_column_in_block_size_bytes, column_names, part.ranges, use_uncompressed_cache,
+                    prewhere_actions, prewhere_column, true, settings.min_bytes_to_use_direct_io, settings.max_read_buffer_size, true,
+                    virt_columns, part.part_index_in_query);
+                // source_stream = std::make_shared<DebugPrintBlockInputStream>(source_stream,
+                //     "partition-" + part.data_part->partition.getID(data) + ".part-" + part.data_part->name);
+                source_stream = std::make_shared<ExpressionBlockInputStream>(source_stream, data.getPrimaryExpression());
+                source_stream = std::make_shared<AsynchronousBlockInputStream>(source_stream);
+                merging.emplace_back(source_stream);
+            }
+
+            // TODO: Optimization: if (merging.size() == 1)
+            BlockInputStreamPtr merged = std::make_shared<ReplacingDeletingSortedBlockInputStream>(merging, data.getSortDescription(),
+                MutableSupport::version_column_name, MutableSupport::delmark_column_name, DEFAULT_MERGE_BLOCK_SIZE, nullptr, false, true);
+
+            // merged = std::make_shared<DebugPrintBlockInputStream>(merged, "partition-" + parts[0].data_part->partition.getID(data));
+            // merged = std::make_shared<AsynchronousBlockInputStream>(merged);
+            res.emplace_back(merged);
+        }
+    }
+    else if (MutableSupport::deduper == MutableSupport::DeduperDedupPartitioning)
+    {
+        for (auto it = partitions.begin(); it != partitions.end(); ++it)
+        {
+            RangesInDataParts & parts = it->second;
+            BlockInputStreams merging;
+
+            for (size_t i = 0; i < parts.size(); ++i)
+            {
+                RangesInDataPart & part = parts[i];
+
+                BlockInputStreamPtr source_stream = std::make_shared<MergeTreeBlockInputStream>(
+                    data, part.data_part, max_block_size, settings.preferred_block_size_bytes,
+                    settings.preferred_max_column_in_block_size_bytes, column_names, part.ranges, use_uncompressed_cache,
+                    prewhere_actions, prewhere_column, true, settings.min_bytes_to_use_direct_io, settings.max_read_buffer_size, true,
+                    virt_columns, part.part_index_in_query);
+                // source_stream = std::make_shared<DebugPrintBlockInputStream>(source_stream,
+                //     "partition-" + part.data_part->partition.getID(data) + ".part-" + part.data_part->name);
+                source_stream = std::make_shared<ExpressionBlockInputStream>(source_stream, data.getPrimaryExpression());
+                merging.emplace_back(source_stream);
+            }
+
+            BlockInputStreamPtr merged = std::make_shared<DedupSortedBlockInputStream>(merging, data.getSortDescription());
+            // merged = std::make_shared<DebugPrintBlockInputStream>(merged, "partition-" + parts[0].data_part->partition.getID(data));
+            res.emplace_back(merged);
+        }
+    }
+
+    return res;
+}
+
+
 void MergeTreeDataSelectExecutor::createPositiveSignCondition(
     ExpressionActionsPtr & out_expression, String & out_column, const Context & context) const
 {
@@ -948,4 +1144,45 @@ MarkRanges MergeTreeDataSelectExecutor::markRangesFromPKRange(
     return res;
 }
 
+
+void MergeTreeDataSelectExecutor::getPartsWithRanges(MergeTreeData::DataPartsVector & parts, const PKCondition & key_condition,
+    const Settings & settings, RangesInDataParts & parts_with_ranges, size_t * out_sum_marks) const
+{
+    size_t part_index = 0;
+    size_t sum_marks = 0;
+    size_t sum_ranges = 0;
+
+    for (auto & part : parts)
+    {
+        RangesInDataPart ranges(part, part_index++);
+
+        if (data.merging_params.mode != MergeTreeData::MergingParams::Unsorted)
+            ranges.ranges = markRangesFromPKRange(part->index, key_condition, settings);
+        else
+            ranges.ranges = MarkRanges{MarkRange{0, part->marks_count}};
+
+        if (!ranges.ranges.empty())
+        {
+            parts_with_ranges.push_back(ranges);
+
+            sum_ranges += ranges.ranges.size();
+            for (const auto & range : ranges.ranges)
+                sum_marks += range.end - range.begin;
+        }
+    }
+
+    if (out_sum_marks)
+        *out_sum_marks = sum_marks;
+
+    LOG_DEBUG(log, "Selected " << parts.size() << " parts by date, " << parts_with_ranges.size() << " parts by key, "
+        << sum_marks << " marks to read from " << sum_ranges << " ranges");
+
+    if (parts_with_ranges.empty())
+        return;
+
+    ProfileEvents::increment(ProfileEvents::SelectedParts, parts_with_ranges.size());
+    ProfileEvents::increment(ProfileEvents::SelectedRanges, sum_ranges);
+    ProfileEvents::increment(ProfileEvents::SelectedMarks, sum_marks);
+}
+
 }
