diff --git a/dbms/src/Storages/MergeTree/MergeTreeData.cpp b/dbms/src/Storages/MergeTree/MergeTreeData.cpp
index 5f9a7a0..7c66775 100644
--- a/dbms/src/Storages/MergeTree/MergeTreeData.cpp
+++ b/dbms/src/Storages/MergeTree/MergeTreeData.cpp
@@ -313,8 +313,8 @@ void MergeTreeData::MergingParams::check(const NamesAndTypesList & columns) cons
     /// Check that version_column column is set only for Replacing mode and is of unsigned integer type.
     if (!version_column.empty())
     {
-        if (mode != MergingParams::Replacing)
-            throw Exception("Version column for MergeTree cannot be specified in all modes except Replacing.",
+        if (mode != MergingParams::Replacing && mode != MergingParams::Mutable)
+            throw Exception("Version column for MergeTree cannot be specified in all modes except Replacing or Mutable.",
                 ErrorCodes::LOGICAL_ERROR);
 
         bool miss_column = true;
@@ -353,6 +353,7 @@ String MergeTreeData::MergingParams::getModeName() const
         case Aggregating:   return "Aggregating";
         case Unsorted:      return "Unsorted";
         case Replacing:     return "Replacing";
+        case Mutable:       return "Mutable";
         case Graphite:      return "Graphite";
 
         default:
