diff --git a/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.h b/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.h
index fd0b026..e91768e 100644
--- a/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.h
+++ b/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.h
@@ -3,6 +3,7 @@
 #include <Storages/MergeTree/MergeTreeData.h>
 #include <Storages/MergeTree/RangesInDataPart.h>
 
+#include <Parsers/ASTSelectQuery.h>
 
 namespace DB
 {
@@ -40,6 +41,7 @@ private:
         const Names & column_names,
         size_t max_block_size,
         bool use_uncompressed_cache,
+        ASTSelectQuery & select,
         ExpressionActionsPtr prewhere_actions,
         const String & prewhere_column,
         const Names & virt_columns,
@@ -50,6 +52,7 @@ private:
         const Names & column_names,
         size_t max_block_size,
         bool use_uncompressed_cache,
+        ASTSelectQuery & select,
         ExpressionActionsPtr prewhere_actions,
         const String & prewhere_column,
         const Names & virt_columns,
