diff --git a/dbms/src/Storages/MergeTree/MergeTreePartition.cpp b/dbms/src/Storages/MergeTree/MergeTreePartition.cpp
index 56de34f..2db26ac 100644
--- a/dbms/src/Storages/MergeTree/MergeTreePartition.cpp
+++ b/dbms/src/Storages/MergeTree/MergeTreePartition.cpp
@@ -23,7 +23,7 @@ static ReadBufferFromFile openForReading(const String & path)
 /// So if you want to change this method, be sure to guarantee compatibility with existing table data.
 String MergeTreePartition::getID(const MergeTreeData & storage) const
 {
-    if (value.size() != storage.partition_expr_columns.size())
+    if (value.size() != storage.partition_expr_columns.size() && storage.merging_params.mode != MergeTreeData::MergingParams::Mutable)
         throw Exception("Invalid partition key size: " + toString(value.size()), ErrorCodes::LOGICAL_ERROR);
 
     if (value.empty())
