diff --git a/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp b/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp
index 1766f60..c22d542 100644
--- a/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp
+++ b/dbms/src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp
@@ -7,10 +7,14 @@
 #include <Storages/MergeTree/MergeTreeReadPool.h>
 #include <Storages/MergeTree/MergeTreeThreadBlockInputStream.h>
 #include <Storages/MergeTree/PKCondition.h>
+
 #include <Parsers/ASTIdentifier.h>
 #include <Parsers/ASTFunction.h>
 #include <Parsers/ASTSampleRatio.h>
 
+#include <Common/Stopwatch.h>
+
+
 /// Allow to use __uint128_t as a template parameter for boost::rational.
 // https://stackoverflow.com/questions/41198673/uint128-t-not-working-with-clang-and-libstdc
 #if !defined(__GLIBCXX_BITSIZE_INT_N_0) && defined(__SIZEOF_INT128__)
@@ -43,6 +47,11 @@ namespace std
 #include <DataTypes/DataTypeEnum.h>
 #include <Storages/VirtualColumnUtils.h>
 
+#include <Storages/MutableSupport.h>
+#include <DataStreams/ExtraHeadTailBlockInputStream.h>
+#include <DataStreams/ReplacingDeletingSortedBlockInputStream.h>
+#include <DataStreams/AsynchronousBlockInputStream.h>
+
 
 namespace ProfileEvents
 {
@@ -138,8 +147,6 @@ BlockInputStreams MergeTreeDataSelectExecutor::read(
     const unsigned num_streams,
     Int64 max_block_number_to_read) const
 {
-    size_t part_index = 0;
-
     MergeTreeData::DataPartsVector parts = data.getDataPartsVector();
 
     /// If query contains restrictions on the virtual column `_part` or `_part_index`, select only parts suitable for it.
@@ -510,43 +517,45 @@ BlockInputStreams MergeTreeDataSelectExecutor::read(
             CreatingSetsBlockInputStream(std::make_shared<NullBlockInputStream>(), prewhere_subqueries, settings.limits).read();
     }
 
-    RangesInDataParts parts_with_ranges;
+    bool is_mutable_engine = (data.merging_params.mode == MergeTreeData::MergingParams::Mutable && !select.raw_for_mutable);
 
-    /// Let's find what range to read from each part.
-    size_t sum_marks = 0;
-    size_t sum_ranges = 0;
-    for (auto & part : parts)
+    RangesInDataParts parts_with_ranges;
+    if (!is_mutable_engine)
     {
-        RangesInDataPart ranges(part, part_index++);
-
-        if (data.merging_params.mode != MergeTreeData::MergingParams::Unsorted)
-            ranges.ranges = markRangesFromPKRange(part->index, key_condition, settings);
-        else
-            ranges.ranges = MarkRanges{MarkRange{0, part->marks_count}};
-
-        if (!ranges.ranges.empty())
-        {
-            parts_with_ranges.push_back(ranges);
-
-            sum_ranges += ranges.ranges.size();
-            for (const auto & range : ranges.ranges)
-                sum_marks += range.end - range.begin;
-        }
+        getPartsWithRanges(parts, key_condition, settings, parts_with_ranges);
+        if (parts_with_ranges.empty())
+            return {};
     }
 
-    LOG_DEBUG(log, "Selected " << parts.size() << " parts by date, " << parts_with_ranges.size() << " parts by key, "
-        << sum_marks << " marks to read from " << sum_ranges << " ranges");
+    BlockInputStreams res;
 
-    if (parts_with_ranges.empty())
-        return {};
+    if (is_mutable_engine)
+    {
+        column_names_to_read.insert(column_names_to_read.end(), MutableSupport::version_column_name);
+        column_names_to_read.insert(column_names_to_read.end(), MutableSupport::delmark_column_name);
 
-    ProfileEvents::increment(ProfileEvents::SelectedParts, parts_with_ranges.size());
-    ProfileEvents::increment(ProfileEvents::SelectedRanges, sum_ranges);
-    ProfileEvents::increment(ProfileEvents::SelectedMarks, sum_marks);
+        std::vector<String> add_columns = data.getPrimaryExpression()->getRequiredColumns();
+        column_names_to_read.insert(column_names_to_read.end(), add_columns.begin(), add_columns.end());
 
-    BlockInputStreams res;
+        std::sort(column_names_to_read.begin(), column_names_to_read.end());
+        column_names_to_read.erase(std::unique(column_names_to_read.begin(), column_names_to_read.end()), column_names_to_read.end());
+
+        res = spreadMarkRangesAmongStreamsOnMutableEngine(
+            parts,
+            key_condition,
+            column_names_to_read,
+            max_block_size,
+            num_streams,
+            settings.use_uncompressed_cache,
+            prewhere_actions,
+            prewhere_column,
+            virt_column_names,
+            settings);
 
-    if (select.final())
+        if (res.size() == 0)
+            return res;
+    }
+    else if (select.final())
     {
         /// Add columns needed to calculate primary key and the sign.
         std::vector<String> add_columns = data.getPrimaryExpression()->getRequiredColumns();
@@ -741,6 +750,7 @@ BlockInputStreams MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreams(
     return res;
 }
 
+
 BlockInputStreams MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreamsFinal(
     RangesInDataParts && parts,
     const Names & column_names,
@@ -823,6 +833,9 @@ BlockInputStreams MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreamsFinal
                     data.getSortDescription(), data.merging_params.version_column, max_block_size);
                 break;
 
+            case MergeTreeData::MergingParams::Mutable:
+                throw Exception("MutableMergeTree doesn't handle here", ErrorCodes::LOGICAL_ERROR);
+
             case MergeTreeData::MergingParams::Unsorted:
                 throw Exception("UnsortedMergeTree doesn't support FINAL", ErrorCodes::LOGICAL_ERROR);
 
@@ -837,6 +850,122 @@ BlockInputStreams MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreamsFinal
 }
 
 
+BlockInputStreams MergeTreeDataSelectExecutor::spreadMarkRangesAmongStreamsOnMutableEngine(
+    MergeTreeData::DataPartsVector & parts,
+    const PKCondition & key_condition,
+    const Names & column_names,
+    size_t max_block_size,
+    unsigned num_streams,
+    bool use_uncompressed_cache,
+    ExpressionActionsPtr prewhere_actions,
+    const String & prewhere_column,
+    const Names & virt_columns,
+    const Settings & settings) const
+{
+    LOG_DEBUG(log, "Number of streams: " << num_streams);
+
+    RangesInDataParts parts_with_ranges;
+    size_t sum_marks = 0;
+    getPartsWithRanges(parts, key_condition, settings, parts_with_ranges, &sum_marks);
+
+    const size_t max_marks_to_use_cache =
+        (settings.merge_tree_max_rows_to_use_cache + data.index_granularity - 1) / data.index_granularity;
+    if (sum_marks > max_marks_to_use_cache)
+        use_uncompressed_cache = false;
+
+    std::unordered_map<std::string, RangesInDataParts> partitions;
+    for (size_t part_index = 0; part_index < parts_with_ranges.size(); ++part_index)
+    {
+        RangesInDataPart & part = parts_with_ranges[part_index];
+        auto partition_id = part.data_part->partition.getID(data);
+        if (partitions.find(partition_id) == partitions.end())
+            partitions[partition_id] = RangesInDataParts();
+        partitions[partition_id].emplace_back(part);
+    }
+
+    // TODO: Use one stream if data is small enough
+    bool origin_streams = false;
+    bool origin_unity = false;
+    bool unity_stream = true;
+
+    if (origin_unity)
+    {
+        BlockInputStreams res;
+        BlockInputStreams merging;
+        for (auto it = parts_with_ranges.begin(); it != parts_with_ranges.end(); ++it)
+        {
+            RangesInDataPart & part = *it;
+
+            BlockInputStreamPtr source_stream = std::make_shared<MergeTreeBlockInputStream>(
+                data, part.data_part, max_block_size, settings.preferred_block_size_bytes,
+                settings.preferred_max_column_in_block_size_bytes, column_names, part.ranges, use_uncompressed_cache,
+                prewhere_actions, prewhere_column, true, settings.min_bytes_to_use_direct_io, settings.max_read_buffer_size, true,
+                virt_columns, part.part_index_in_query);
+
+            merging.emplace_back(std::make_shared<ExpressionBlockInputStream>(source_stream, data.getPrimaryExpression()));
+        }
+        if (origin_streams)
+            return merging;
+        res.emplace_back(std::make_shared<ReplacingDeletingSortedBlockInputStream>(merging, data.getSortDescription(),
+            MutableSupport::version_column_name, MutableSupport::delmark_column_name, DEFAULT_MERGE_BLOCK_SIZE, nullptr, false, true));
+        return res;
+    }
+    else if (unity_stream)
+    {
+        BlockInputStreams res;
+        BlockInputStreams merging;
+        for (auto it = partitions.begin(); it != partitions.end(); ++it)
+        {
+            RangesInDataParts & parts = it->second;
+
+            for (size_t i = 0; i < parts.size(); ++i)
+            {
+                RangesInDataPart & part = parts[i];
+
+                BlockInputStreamPtr source_stream = std::make_shared<MergeTreeBlockInputStream>(
+                    data, part.data_part, max_block_size, settings.preferred_block_size_bytes,
+                    settings.preferred_max_column_in_block_size_bytes, column_names, part.ranges, use_uncompressed_cache,
+                    prewhere_actions, prewhere_column, true, settings.min_bytes_to_use_direct_io, settings.max_read_buffer_size, true,
+                    virt_columns, part.part_index_in_query);
+
+                merging.emplace_back(std::make_shared<ExpressionBlockInputStream>(source_stream, data.getPrimaryExpression()));
+            }
+        }
+        res.emplace_back(std::make_shared<ReplacingDeletingSortedBlockInputStream>(merging, data.getSortDescription(),
+            MutableSupport::version_column_name, MutableSupport::delmark_column_name, DEFAULT_MERGE_BLOCK_SIZE, nullptr, false, true));
+        return res;
+    }
+    else
+    {
+        BlockInputStreams res;
+        for (auto it = partitions.begin(); it != partitions.end(); ++it)
+        {
+            RangesInDataParts & parts = it->second;
+            BlockInputStreams merging;
+
+            LOG_DEBUG(log, "PROBE Partition: " << parts[0].data_part->partition.getID(data));
+            for (size_t i = 0; i < parts.size(); ++i)
+            {
+                RangesInDataPart & part = parts[i];
+                LOG_DEBUG(log, "PROBE Partition: " << part.data_part->name);
+
+                BlockInputStreamPtr source_stream = std::make_shared<MergeTreeBlockInputStream>(
+                    data, part.data_part, max_block_size, settings.preferred_block_size_bytes,
+                    settings.preferred_max_column_in_block_size_bytes, column_names, part.ranges, use_uncompressed_cache,
+                    prewhere_actions, prewhere_column, true, settings.min_bytes_to_use_direct_io, settings.max_read_buffer_size, true,
+                    virt_columns, part.part_index_in_query);
+
+                merging.emplace_back(std::make_shared<ExpressionBlockInputStream>(source_stream, data.getPrimaryExpression()));
+            }
+
+            res.emplace_back(std::make_shared<ReplacingDeletingSortedBlockInputStream>(merging, data.getSortDescription(),
+                MutableSupport::version_column_name, MutableSupport::delmark_column_name, DEFAULT_MERGE_BLOCK_SIZE, nullptr, false, true));
+        }
+        return res;
+    }
+}
+
+
 void MergeTreeDataSelectExecutor::createPositiveSignCondition(
     ExpressionActionsPtr & out_expression, String & out_column, const Context & context) const
 {
@@ -948,4 +1077,45 @@ MarkRanges MergeTreeDataSelectExecutor::markRangesFromPKRange(
     return res;
 }
 
+
+void MergeTreeDataSelectExecutor::getPartsWithRanges(MergeTreeData::DataPartsVector & parts, const PKCondition & key_condition,
+    const Settings & settings, RangesInDataParts & parts_with_ranges, size_t * out_sum_marks) const
+{
+    size_t part_index = 0;
+    size_t sum_marks = 0;
+    size_t sum_ranges = 0;
+
+    for (auto & part : parts)
+    {
+        RangesInDataPart ranges(part, part_index++);
+
+        if (data.merging_params.mode != MergeTreeData::MergingParams::Unsorted)
+            ranges.ranges = markRangesFromPKRange(part->index, key_condition, settings);
+        else
+            ranges.ranges = MarkRanges{MarkRange{0, part->marks_count}};
+
+        if (!ranges.ranges.empty())
+        {
+            parts_with_ranges.push_back(ranges);
+
+            sum_ranges += ranges.ranges.size();
+            for (const auto & range : ranges.ranges)
+                sum_marks += range.end - range.begin;
+        }
+    }
+
+    if (out_sum_marks)
+        *out_sum_marks = sum_marks;
+
+    LOG_DEBUG(log, "Selected " << parts.size() << " parts by date, " << parts_with_ranges.size() << " parts by key, "
+        << sum_marks << " marks to read from " << sum_ranges << " ranges");
+
+    if (parts_with_ranges.empty())
+        return;
+
+    ProfileEvents::increment(ProfileEvents::SelectedParts, parts_with_ranges.size());
+    ProfileEvents::increment(ProfileEvents::SelectedRanges, sum_ranges);
+    ProfileEvents::increment(ProfileEvents::SelectedMarks, sum_marks);
+}
+
 }
