diff --git a/dbms/src/Server/Server.cpp b/dbms/src/Server/Server.cpp
index 0fe39f4..7d4b10c 100644
--- a/dbms/src/Server/Server.cpp
+++ b/dbms/src/Server/Server.cpp
@@ -41,6 +41,8 @@
 #include <Poco/Net/SecureServerSocket.h>
 #endif
 
+#include "Magic/TCPArrowHandlerFactory.h"
+
 namespace CurrentMetrics
 {
     extern const Metric Revision;
@@ -393,6 +395,23 @@ int Server::main(const std::vector<std::string> & /*args*/)
 #endif
                 }
 
+                /// Arrow via TCP
+                if (config().has("tcp_arrow_port"))
+                {
+                    std::call_once(ssl_init_once, SSLInit);
+                    Poco::Net::SocketAddress tcp_address = make_socket_address(listen_host, config().getInt("tcp_arrow_port"));
+                    Poco::Net::ServerSocket tcp_socket(tcp_address);
+                    tcp_socket.setReceiveTimeout(settings.receive_timeout);
+                    tcp_socket.setSendTimeout(settings.send_timeout);
+                    servers.emplace_back(new Poco::Net::TCPServer(
+                        new Magic::TCPArrowHandlerFactory(*this),
+                        server_pool,
+                        tcp_socket,
+                        new Poco::Net::TCPServerParams));
+
+                    LOG_INFO(log, "Listening tcp(arrow): " + tcp_address.toString());
+                }
+
                 /// TCP
                 if (config().has("tcp_port"))
                 {
