diff --git a/dbms/src/Interpreters/Context.h b/dbms/src/Interpreters/Context.h
index a8bfc5975de6..72370e80b2d2 100644
--- a/dbms/src/Interpreters/Context.h
+++ b/dbms/src/Interpreters/Context.h
@@ -69,7 +69,8 @@ using BlockOutputStreamPtr = std::shared_ptr<IBlockOutputStream>;
 class Block;
 struct SystemLogs;
 using SystemLogsPtr = std::shared_ptr<SystemLogs>;
-
+class SharedQueries;
+using SharedQueriesPtr = std::shared_ptr<SharedQueries>;
 
 /// (database name, table name)
 using DatabaseAndTableName = std::pair<String, String>;
@@ -388,6 +389,8 @@ public:
     String getFormatSchemaPath() const;
     void setFormatSchemaPath(const String & path);
 
+    SharedQueriesPtr getSharedQueries();
+
     /// User name and session identifier. Named sessions are local to users.
     using SessionKey = std::pair<String, String>;
 
