diff --git a/dbms/src/Functions/FunctionsRound.h b/dbms/src/Functions/FunctionsRound.h
index d5429b76318e..6cb3244aae6c 100644
--- a/dbms/src/Functions/FunctionsRound.h
+++ b/dbms/src/Functions/FunctionsRound.h
@@ -73,6 +73,12 @@ roundDownToPowerOfTwo(T x)
     return ext::bit_cast<T>(ext::bit_cast<UInt64>(x) & ~((1ULL << 52) - 1));
 }
 
+template <typename T>
+inline std::enable_if_t<std::is_same_v<DecimalValue, T>, T> roundDownToPowerOfTwo(T)
+{
+    return {};
+}
+
 /** For integer data types:
   * - if number is greater than zero, round it down to nearest power of two (example: roundToExp2(100) = 64, roundToExp2(64) = 64);
   * - otherwise, return 0.
