diff --git a/dbms/src/Storages/MergeTree/MergedBlockOutputStream.h b/dbms/src/Storages/MergeTree/MergedBlockOutputStream.h
index 4b83f9599918..57e211df427a 100644
--- a/dbms/src/Storages/MergeTree/MergedBlockOutputStream.h
+++ b/dbms/src/Storages/MergeTree/MergedBlockOutputStream.h
@@ -1,8 +1,10 @@
 #pragma once
 
 #include <IO/WriteBufferFromFile.h>
+#include <IO/WriteBufferFromOStream.h>
 #include <IO/CompressedWriteBuffer.h>
 #include <IO/HashingWriteBuffer.h>
+#include <IO/CompactContext.h>
 #include <Storages/MergeTree/MergeTreeData.h>
 #include <DataStreams/IBlockOutputStream.h>
 
@@ -39,18 +41,24 @@ protected:
             size_t estimated_size,
             size_t aio_threshold);
 
+        ColumnStream(
+            const String & escaped_column_name_,
+            CompressionSettings compression_settings,
+            const CompactWriteContextPtr& compactCtxPtr);
+
         String escaped_column_name;
         std::string data_file_extension;
         std::string marks_file_extension;
 
         /// compressed -> compressed_buf -> plain_hashing -> plain_file
         std::unique_ptr<WriteBufferFromFileBase> plain_file;
-        HashingWriteBuffer plain_hashing;
+        HashingWriteBuffer *plain_hashing;
         CompressedWriteBuffer compressed_buf;
         HashingWriteBuffer compressed;
 
         /// marks -> marks_file
-        WriteBufferFromFile marks_file;
+        std::unique_ptr<WriteBufferFromFile> marks_file;
+        std::unique_ptr<WriteBufferFromOStream> marks_sstream;
         HashingWriteBuffer marks;
 
         void finalize();
@@ -80,6 +88,8 @@ protected:
     size_t aio_threshold;
 
     CompressionSettings compression_settings;
+
+    CompactWriteContextPtr compactCtxPtr;
 };
 
 
@@ -93,12 +103,14 @@ public:
         MergeTreeData & storage_,
         String part_path_,
         const NamesAndTypesList & columns_list_,
+        bool use_compact_write,
         CompressionSettings compression_settings);
 
     MergedBlockOutputStream(
         MergeTreeData & storage_,
         String part_path_,
         const NamesAndTypesList & columns_list_,
+        bool use_compact_write,
         CompressionSettings compression_settings,
         const MergeTreeData::DataPart::ColumnToSize & merged_column_to_size_,
         size_t aio_threshold_);
