diff --git a/dbms/src/Storages/MergeTree/MergeTreeBlockInputStream.h b/dbms/src/Storages/MergeTree/MergeTreeBlockInputStream.h
index 13bd55ba7ebe..a5503da92b24 100644
--- a/dbms/src/Storages/MergeTree/MergeTreeBlockInputStream.h
+++ b/dbms/src/Storages/MergeTree/MergeTreeBlockInputStream.h
@@ -30,6 +30,7 @@ public:
         size_t min_bytes_to_use_direct_io,
         size_t max_read_buffer_size,
         bool save_marks_in_cache,
+        bool update_persisted_cache,
         const Names & virt_column_names = {},
         size_t part_index_in_query = 0,
         bool quiet = false);
