diff --git a/dbms/src/Storages/MergeTree/MergeTreeBlockOutputStream.h b/dbms/src/Storages/MergeTree/MergeTreeBlockOutputStream.h
index 64243b6e7bf6..bbd373819931 100644
--- a/dbms/src/Storages/MergeTree/MergeTreeBlockOutputStream.h
+++ b/dbms/src/Storages/MergeTree/MergeTreeBlockOutputStream.h
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <DataStreams/IBlockOutputStream.h>
+#include <common/logger_useful.h>
 
 
 namespace DB
@@ -14,13 +15,14 @@ class MergeTreeBlockOutputStream : public IBlockOutputStream
 {
 public:
     MergeTreeBlockOutputStream(StorageMergeTree & storage_)
-        : storage(storage_) {}
+        : storage(storage_), log(&Logger::get("MergeTreeBlockOutputStream")) {}
 
     Block getHeader() const override;
     void write(const Block & block) override;
 
 private:
     StorageMergeTree & storage;
+    Logger *log;
 };
 
 }
