diff --git a/dbms/src/DataTypes/DataTypeDate.h b/dbms/src/DataTypes/DataTypeDate.h
index cef5d9ae2e04..6af0ebfedf18 100644
--- a/dbms/src/DataTypes/DataTypeDate.h
+++ b/dbms/src/DataTypes/DataTypeDate.h
@@ -6,7 +6,7 @@
 namespace DB
 {
 
-class DataTypeDate final : public DataTypeNumberBase<UInt16>
+class DataTypeDate final : public DataTypeNumberBase<UInt32>
 {
 public:
     const char * getFamilyName() const override { return "Date"; }
