diff --git a/dbms/src/DataTypes/IDataType.cpp b/dbms/src/DataTypes/IDataType.cpp
index 68fe74d18f05..0edcebb7f3c1 100644
--- a/dbms/src/DataTypes/IDataType.cpp
+++ b/dbms/src/DataTypes/IDataType.cpp
@@ -68,6 +68,15 @@ size_t IDataType::getSizeOfValueInMemory() const
 }
 
 
+bool IDataType::isNullMap(const IDataType::SubstreamPath & path) {
+    for (const Substream & elem : path)
+    {
+        if (elem.type == Substream::NullMap)
+            return true;
+    }
+    return false;
+}
+
 String IDataType::getFileNameForStream(const String & column_name, const IDataType::SubstreamPath & path)
 {
     String nested_table_name = Nested::extractTableName(column_name);
