diff --git a/dbms/src/Parsers/ASTSelectQuery.h b/dbms/src/Parsers/ASTSelectQuery.h
index c31e91f31949..580d248d4559 100644
--- a/dbms/src/Parsers/ASTSelectQuery.h
+++ b/dbms/src/Parsers/ASTSelectQuery.h
@@ -20,6 +20,7 @@ public:
 
     ASTPtr clone() const override;
 
+    bool raw_for_mutable = false;
     bool distinct = false;
     ASTPtr with_expression_list;
     ASTPtr select_expression_list;
