diff --git a/dbms/src/Parsers/ASTSelectQuery.cpp b/dbms/src/Parsers/ASTSelectQuery.cpp
index f234b0a..e2938d4 100644
--- a/dbms/src/Parsers/ASTSelectQuery.cpp
+++ b/dbms/src/Parsers/ASTSelectQuery.cpp
@@ -71,7 +71,12 @@ void ASTSelectQuery::formatImpl(const FormatSettings & s, FormatState & state, F
         s.ostr << s.nl_or_ws;
     }
 
-    s.ostr << (s.hilite ? hilite_keyword : "") << indent_str << "SELECT " << (distinct ? "DISTINCT " : "") << (s.hilite ? hilite_none : "");
+    s.ostr
+        << (s.hilite ? hilite_keyword : "")
+        << indent_str
+        << (raw_for_mutable ? "SELRAW " : "SELECT ")
+        << (distinct ? "DISTINCT " : "")
+        << (s.hilite ? hilite_none : "");
 
     s.one_line
         ? select_expression_list->formatImpl(s, state, frame)
