diff --git a/dbms/src/Parsers/ASTSelectQuery.cpp b/dbms/src/Parsers/ASTSelectQuery.cpp
index 136a5ed..527f921 100644
--- a/dbms/src/Parsers/ASTSelectQuery.cpp
+++ b/dbms/src/Parsers/ASTSelectQuery.cpp
@@ -248,7 +248,12 @@ void ASTSelectQuery::formatQueryImpl(const FormatSettings & s, FormatState & sta
         s.ostr << s.nl_or_ws;
     }
 
-    s.ostr << (s.hilite ? hilite_keyword : "") << indent_str << "SELECT " << (distinct ? "DISTINCT " : "") << (s.hilite ? hilite_none : "");
+    s.ostr
+        << (s.hilite ? hilite_keyword : "")
+        << indent_str
+        << (raw_for_mutable ? "SELRAW " : "SELECT ")
+        << (distinct ? "DISTINCT " : "")
+        << (s.hilite ? hilite_none : "");
 
     s.one_line
         ? select_expression_list->formatImpl(s, state, frame)
