diff --git a/dbms/src/Server/Client.cpp b/dbms/src/Server/Client.cpp
index 362fdc2f6570..f2acafa75ac5 100644
--- a/dbms/src/Server/Client.cpp
+++ b/dbms/src/Server/Client.cpp
@@ -1503,6 +1503,7 @@ public:
             max_client_network_bandwidth = options["max_client_network_bandwidth"].as<int>();
         if (options.count("compression"))
             config().setBool("compression", options["compression"].as<bool>());
+
     }
 };
 
