diff --git a/dbms/src/Server/Server.cpp b/dbms/src/Server/Server.cpp
index 66afef5cfc8f..e7b9ebb6cc41 100644
--- a/dbms/src/Server/Server.cpp
+++ b/dbms/src/Server/Server.cpp
@@ -276,6 +276,16 @@ int Server::main(const std::vector<std::string> & /*args*/)
     if (uncompressed_cache_size)
         global_context->setUncompressedCache(uncompressed_cache_size);
 
+    /// Quota size in bytes of persisted mapping cache. 0 means unlimited.
+    size_t persisted_cache_size = config().getUInt64("persisted_mapping_cache_size", 0);
+    /// Path of persisted cache in fast(er) disk device. Empty means disabled.
+    std::string persisted_cache_path = config().getString("persisted_mapping_cache_path", "");
+    if (!persisted_cache_path.empty())
+        global_context->setPersistedCache(persisted_cache_size, persisted_cache_path);
+
+    bool use_L0_opt = config().getBool("l0_optimize", true);
+    global_context->setUseL0Opt(use_L0_opt);
+
     /// Load global settings from default_profile and system_profile.
     global_context->setDefaultProfiles(config());
     Settings & settings = global_context->getSettingsRef();
