diff --git a/dbms/src/Server/TCPHandler.h b/dbms/src/Server/TCPHandler.h
index 7202067dab8b..51a77541d52a 100644
--- a/dbms/src/Server/TCPHandler.h
+++ b/dbms/src/Server/TCPHandler.h
@@ -134,7 +134,7 @@ private:
 
     /// Process a request that does not require the receiving of data blocks from the client
     void processOrdinaryQuery();
-
+    void processSharedQuery();
     void processTablesStatusRequest();
 
     void sendHello();
