diff --git a/dbms/src/Storages/IStorage.h b/dbms/src/Storages/IStorage.h
index 74db06041471..579822f88e3d 100644
--- a/dbms/src/Storages/IStorage.h
+++ b/dbms/src/Storages/IStorage.h
@@ -231,6 +231,13 @@ public:
         throw Exception("Method dropPartition is not supported by storage " + getName(), ErrorCodes::NOT_IMPLEMENTED);
     }
 
+    /** Run the query TRUNCATE TABLE.
+      */
+    virtual void truncate(const ASTPtr & /*query*/, const Context & /*context*/)
+    {
+      throw Exception("Method truncate is not supported by storage " + getName(), ErrorCodes::NOT_IMPLEMENTED);
+    }
+
     /** Run the ATTACH request (PART|PARTITION).
       */
     virtual void attachPartition(const ASTPtr & /*partition*/, bool /*part*/, const Context & /*context*/)
