diff --git a/dbms/src/Common/FieldVisitors.cpp b/dbms/src/Common/FieldVisitors.cpp
index 2243a838a991..280ea7e04d50 100644
--- a/dbms/src/Common/FieldVisitors.cpp
+++ b/dbms/src/Common/FieldVisitors.cpp
@@ -34,6 +34,7 @@ String FieldVisitorDump::operator() (const Null &) const { return "NULL"; }
 String FieldVisitorDump::operator() (const UInt64 & x) const { return formatQuotedWithPrefix(x, "UInt64_"); }
 String FieldVisitorDump::operator() (const Int64 & x) const { return formatQuotedWithPrefix(x, "Int64_"); }
 String FieldVisitorDump::operator() (const Float64 & x) const { return formatQuotedWithPrefix(x, "Float64_"); }
+String FieldVisitorDump::operator() (const Decimal & x) const { return "Decimal_" + x.toString(); }
 
 
 String FieldVisitorDump::operator() (const String & x) const
@@ -104,6 +105,7 @@ String FieldVisitorToString::operator() (const UInt64 & x) const { return format
 String FieldVisitorToString::operator() (const Int64 & x) const { return formatQuoted(x); }
 String FieldVisitorToString::operator() (const Float64 & x) const { return formatFloat(x); }
 String FieldVisitorToString::operator() (const String & x) const { return formatQuoted(x); }
+String FieldVisitorToString::operator() (const Decimal & x) const { return "Decimal_" + x.toString(); }
 
 
 String FieldVisitorToString::operator() (const Array & x) const
@@ -169,6 +171,14 @@ void FieldVisitorHash::operator() (const Float64 & x) const
     hash.update(x);
 }
 
+void FieldVisitorHash::operator() (const Decimal & x) const
+{
+    UInt8 type = Field::Types::Decimal;
+    hash.update(type);
+    hash.update(x);
+}
+
+
 void FieldVisitorHash::operator() (const String & x) const
 {
     UInt8 type = Field::Types::String;
