diff --git a/dbms/src/Dictionaries/MongoDBBlockInputStream.cpp b/dbms/src/Dictionaries/MongoDBBlockInputStream.cpp
index b5f1d1f6f697..11125fb08ed0 100644
--- a/dbms/src/Dictionaries/MongoDBBlockInputStream.cpp
+++ b/dbms/src/Dictionaries/MongoDBBlockInputStream.cpp
@@ -116,8 +116,8 @@ namespace
                         "Type mismatch, expected Timestamp, got type id = " + toString(value.type()) +
                             " for column " + name, ErrorCodes::TYPE_MISMATCH};
 
-                static_cast<ColumnUInt16 &>(column).getData().push_back(
-                    UInt16{DateLUT::instance().toDayNum(
+                static_cast<ColumnUInt32 &>(column).getData().push_back(
+                    UInt32{DateLUT::instance().toDayNum(
                         static_cast<const Poco::MongoDB::ConcreteElement<Poco::Timestamp> &>(value).value().epochTime())});
                 break;
             }
