diff --git a/dbms/src/Dictionaries/MySQLBlockInputStream.cpp b/dbms/src/Dictionaries/MySQLBlockInputStream.cpp
index 6e2275cd1059..a3fc591ef7dc 100644
--- a/dbms/src/Dictionaries/MySQLBlockInputStream.cpp
+++ b/dbms/src/Dictionaries/MySQLBlockInputStream.cpp
@@ -52,7 +52,7 @@ namespace
             case ValueType::Float32: static_cast<ColumnFloat32 &>(column).insert(value.getDouble()); break;
             case ValueType::Float64: static_cast<ColumnFloat64 &>(column).insert(value.getDouble()); break;
             case ValueType::String: static_cast<ColumnString &>(column).insertData(value.data(), value.size()); break;
-            case ValueType::Date: static_cast<ColumnUInt16 &>(column).insert(UInt16{value.getDate().getDayNum()}); break;
+            case ValueType::Date: static_cast<ColumnUInt32 &>(column).insert(UInt32{value.getDate().getDayNum()}); break;
             case ValueType::DateTime: static_cast<ColumnUInt32 &>(column).insert(time_t{value.getDateTime()}); break;
         }
     }
