diff --git a/dbms/src/Dictionaries/ODBCBlockInputStream.cpp b/dbms/src/Dictionaries/ODBCBlockInputStream.cpp
index 8b186f6791ee..d6ba9f5cdd9c 100644
--- a/dbms/src/Dictionaries/ODBCBlockInputStream.cpp
+++ b/dbms/src/Dictionaries/ODBCBlockInputStream.cpp
@@ -57,7 +57,7 @@ namespace
             case ValueType::Float32: static_cast<ColumnFloat32 &>(column).insert(value.convert<Float64>()); break;
             case ValueType::Float64: static_cast<ColumnFloat64 &>(column).insert(value.convert<Float64>()); break;
             case ValueType::String: static_cast<ColumnString &>(column).insert(value.convert<String>()); break;
-            case ValueType::Date: static_cast<ColumnUInt16 &>(column).insert(UInt16{LocalDate{value.convert<String>()}.getDayNum()}); break;
+            case ValueType::Date: static_cast<ColumnUInt32 &>(column).insert(UInt32{LocalDate{value.convert<String>()}.getDayNum()}); break;
             case ValueType::DateTime: static_cast<ColumnUInt32 &>(column).insert(time_t{LocalDateTime{value.convert<String>()}}); break;
         }
     }
