diff --git a/dbms/src/Functions/FunctionsExternalDictionaries.h b/dbms/src/Functions/FunctionsExternalDictionaries.h
index bbf9b1b5c603..4cf242be6dc0 100644
--- a/dbms/src/Functions/FunctionsExternalDictionaries.h
+++ b/dbms/src/Functions/FunctionsExternalDictionaries.h
@@ -694,8 +694,8 @@ DECLARE_DICT_GET_TRAITS(Int32, DataTypeInt32)
 DECLARE_DICT_GET_TRAITS(Int64, DataTypeInt64)
 DECLARE_DICT_GET_TRAITS(Float32, DataTypeFloat32)
 DECLARE_DICT_GET_TRAITS(Float64, DataTypeFloat64)
-DECLARE_DICT_GET_TRAITS(UInt16, DataTypeDate)
-DECLARE_DICT_GET_TRAITS(UInt32, DataTypeDateTime)
+DECLARE_DICT_GET_TRAITS(UInt32, DataTypeDate)
+DECLARE_DICT_GET_TRAITS(UInt64, DataTypeDateTime)
 DECLARE_DICT_GET_TRAITS(UInt128, DataTypeUUID)
 #undef DECLARE_DICT_GET_TRAITS
 
