diff --git a/dbms/src/Storages/ITableDeclaration.h b/dbms/src/Storages/ITableDeclaration.h
index 270ab17..273f92c 100644
--- a/dbms/src/Storages/ITableDeclaration.h
+++ b/dbms/src/Storages/ITableDeclaration.h
@@ -24,6 +24,15 @@ public:
     Block getSampleBlockNonMaterialized() const;
     Block getSampleBlockForColumns(const Names & column_names) const;
 
+    /** The hidden coloumns will not be returned. Mainly for INSERT query.
+      */
+    Block getSampleBlockNoHidden() const;
+    Block getSampleBlockNonMaterializedNoHidden() const;
+
+    NamesAndTypesList getColumnsListNoHidden() const;
+    NamesAndTypesList getColumnsListNonMaterializedNoHidden() const;
+    Names getColumnNamesListNoHidden() const;
+
     /** Verify that all the requested names are in the table and are set correctly.
       * (the list of names is not empty and the names do not repeat)
       */
@@ -50,6 +59,12 @@ public:
 
 private:
     ColumnsDescription columns;
+    OrderedNameSet empty_names;
+
+    virtual const OrderedNameSet & getHiddenColumnsImpl() const
+    {
+        return empty_names;
+    }
 };
 
 }
