diff --git a/dbms/src/Storages/StorageMergeTree.h b/dbms/src/Storages/StorageMergeTree.h
index acb0e03a6..29eed6c96 100644
--- a/dbms/src/Storages/StorageMergeTree.h
+++ b/dbms/src/Storages/StorageMergeTree.h
@@ -59,6 +59,8 @@ public:
 
     BlockOutputStreamPtr write(const ASTPtr & query, const Settings & settings) override;
 
+    UInt64 getVersionSeed(const Settings & settings) const;
+
     /** Perform the next step in combining the parts.
       */
     bool optimize(const ASTPtr & query, const ASTPtr & partition, bool final, bool deduplicate, const Context & context) override;
@@ -113,6 +115,8 @@ private:
 
     BackgroundProcessingPool::TaskHandle merge_task_handle;
 
+    const OrderedNameSet & getHiddenColumnsImpl() const override;
+
     friend struct CurrentlyMergingPartsTagger;
 
     /** Determines what parts should be merged and merges it.
