diff --git a/dbms/src/Common/ErrorCodes.cpp b/dbms/src/Common/ErrorCodes.cpp
index 46b4c3394..49c64dbea 100644
--- a/dbms/src/Common/ErrorCodes.cpp
+++ b/dbms/src/Common/ErrorCodes.cpp
@@ -379,6 +379,11 @@ namespace ErrorCodes
     extern const int POCO_EXCEPTION = 1000;
     extern const int STD_EXCEPTION = 1001;
     extern const int UNKNOWN_EXCEPTION = 1002;
+
+    extern const int THEFLASH_UNHANDLED_TYPE_CONVERTION = 9000;
+    extern const int THEFLASH_BAD_REQUEST = 9001;
+    extern const int THEFLASH_ENCODER_ERROR = 9002;
+    extern const int THEFLASH_SESSION_ERROR = 9003;
 }
 
 }
