diff --git a/dbms/src/Common/SipHash.h b/dbms/src/Common/SipHash.h
index a9f007f7aa43..f30952ba642c 100644
--- a/dbms/src/Common/SipHash.h
+++ b/dbms/src/Common/SipHash.h
@@ -15,6 +15,7 @@
 
 #include <common/Types.h>
 #include <type_traits>
+#include <Common/Decimal.h>
 
 #define ROTL(x, b) static_cast<UInt64>(((x) << (b)) | ((x) >> (64 - (b))))
 
@@ -189,6 +190,13 @@ std::enable_if_t<std::/*has_unique_object_representations_v*/is_standard_layout_
     return hash.get64();
 }
 
+inline UInt64 sipHash64(const DB::DecimalValue& x)
+{
+    SipHash hash;
+    hash.update(reinterpret_cast<const char *>(&x), sizeof(x));
+    return hash.get64();
+}
+
 #include <string>
 
 inline UInt64 sipHash64(const std::string & s)
