diff --git a/dbms/src/Interpreters/InterpreterAlterQuery.cpp b/dbms/src/Interpreters/InterpreterAlterQuery.cpp
index 76fef51..e874649 100644
--- a/dbms/src/Interpreters/InterpreterAlterQuery.cpp
+++ b/dbms/src/Interpreters/InterpreterAlterQuery.cpp
@@ -15,6 +15,9 @@
 #include <Parsers/formatAST.h>
 #include <Parsers/parseQuery.h>
 
+#include <Storages/HiddenColumns.h>
+#include <Storages/StorageMergeTree.h>
+
 #include <Poco/FileStream.h>
 
 #include <algorithm>
@@ -50,7 +53,7 @@ BlockIO InterpreterAlterQuery::execute()
 
     AlterCommands alter_commands;
     PartitionCommands partition_commands;
-    parseAlter(alter.parameters, alter_commands, partition_commands);
+    parseAlter(alter.parameters, alter_commands, partition_commands, table);
 
     partition_commands.validate(table.get());
     for (const PartitionCommand & command : partition_commands)
@@ -90,10 +93,17 @@ BlockIO InterpreterAlterQuery::execute()
 
 void InterpreterAlterQuery::parseAlter(
     const ASTAlterQuery::ParameterContainer & params_container,
-    AlterCommands & out_alter_commands, PartitionCommands & out_partition_commands)
+    AlterCommands & out_alter_commands, PartitionCommands & out_partition_commands, StoragePtr table)
 {
     const DataTypeFactory & data_type_factory = DataTypeFactory::instance();
 
+    OrderedNameSet filtered_names;
+    {
+        StorageMergeTree * merge_tree = typeid_cast<StorageMergeTree *>(&*table);
+        if (merge_tree != nullptr)
+            filtered_names = HiddenColumns::instance().get(merge_tree->getData().merging_params.mode);
+    }
+
     for (const auto & params : params_container)
     {
         if (params.type == ASTAlterQuery::ADD_COLUMN)
@@ -103,6 +113,10 @@ void InterpreterAlterQuery::parseAlter(
 
             const auto & ast_col_decl = typeid_cast<const ASTColumnDeclaration &>(*params.col_decl);
 
+            if (ast_col_decl.name == HiddenColumns::mutable_version_column_name ||
+                ast_col_decl.name == HiddenColumns::mutable_delmark_column_name)
+                throw Exception("Internal column name can not be used: " + ast_col_decl.name, ErrorCodes::ARGUMENT_OUT_OF_BOUND);
+
             command.column_name = ast_col_decl.name;
             if (ast_col_decl.type)
             {
@@ -130,6 +144,13 @@ void InterpreterAlterQuery::parseAlter(
 
                 const Field & column_name = typeid_cast<const ASTIdentifier &>(*(params.column)).name;
 
+                if (column_name == HiddenColumns::mutable_version_column_name ||
+                    column_name == HiddenColumns::mutable_delmark_column_name)
+                {
+                    FieldVisitorToString to_string;
+                    auto err_msg = "Internal column name can not be dropped: " + applyVisitor(to_string, column_name);
+                    throw Exception(err_msg, ErrorCodes::ARGUMENT_OUT_OF_BOUND);
+                }
                 out_partition_commands.emplace_back(PartitionCommand::clearColumn(params.partition, column_name));
             }
             else
@@ -141,6 +162,10 @@ void InterpreterAlterQuery::parseAlter(
                 command.type = AlterCommand::DROP_COLUMN;
                 command.column_name = typeid_cast<const ASTIdentifier &>(*(params.column)).name;
 
+                if (command.column_name == HiddenColumns::mutable_version_column_name ||
+                    command.column_name == HiddenColumns::mutable_delmark_column_name)
+                    throw Exception("Internal column name can not be dropped: " + command.column_name, ErrorCodes::ARGUMENT_OUT_OF_BOUND);
+
                 out_alter_commands.emplace_back(std::move(command));
             }
         }
@@ -151,6 +176,10 @@ void InterpreterAlterQuery::parseAlter(
 
             const auto & ast_col_decl = typeid_cast<const ASTColumnDeclaration &>(*params.col_decl);
 
+            if (ast_col_decl.name == HiddenColumns::mutable_version_column_name ||
+                ast_col_decl.name == HiddenColumns::mutable_delmark_column_name)
+                throw Exception("Internal column name can not be modified: " + ast_col_decl.name, ErrorCodes::ARGUMENT_OUT_OF_BOUND);
+
             command.column_name = ast_col_decl.name;
             if (ast_col_decl.type)
             {
