diff --git a/dbms/src/Interpreters/InterpreterInsertQuery.cpp b/dbms/src/Interpreters/InterpreterInsertQuery.cpp
index 7ae0708..2c7e978 100644
--- a/dbms/src/Interpreters/InterpreterInsertQuery.cpp
+++ b/dbms/src/Interpreters/InterpreterInsertQuery.cpp
@@ -74,9 +74,18 @@ Block InterpreterInsertQuery::getSampleBlock()
 
     /// If the query does not include information about columns
     if (!query.columns)
-        return getTable()->getSampleBlockNonMaterialized();
+    {
+        if (query.is_import)
+            return getTable()->getSampleBlockNonMaterialized();
+        else
+            return getTable()->getSampleBlockNonMaterializedNoHidden();
+    }
 
-    Block table_sample = getTable()->getSampleBlock();
+    Block table_sample;
+    if (query.is_import)
+        table_sample = getTable()->getSampleBlock();
+    else
+        table_sample = getTable()->getSampleBlockNoHidden();
 
     /// Form the block based on the column names from the query
     Block res;
